web/src/main/java/com/epam/esm/web/controller/GiftCertificateController.java

```diff
@@ -0,0 +1,92 @@
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.bind.annotation.PutMapping;
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RequestParam;
+import org.springframework.web.bind.annotation.ResponseStatus;
+import org.springframework.web.bind.annotation.RestController;
+
+import java.util.List;
+import java.util.Optional;
+
+@RestController
+@RequestMapping("/api/certificates")
+public class GiftCertificateController {
+  private GiftCertificateService giftCertificateService;
+
+  @Autowired
+  public GiftCertificateController(GiftCertificateService giftCertificateService) {
+    this.giftCertificateService = giftCertificateService;
+  }
+
+  @GetMapping()
+  public List<GiftCertificateDto> getAll(
+      @RequestParam(value = "name", required = false) String name,
+      @RequestParam(value = "description", required = false) String description,
+      @RequestParam(value = "tag", required = false) String tagName,
+      @RequestParam(value = "sort", required = false) String sort) {
+    if (name == null && description == null && tagName == null) {
+      return giftCertificateService.getAll(sort);
+    }
+    List<GiftCertificateDto> giftCertificateDtos =
+        giftCertificateService.getGiftCertificatesByNameOrDescriptionPart(name, description, sort);
+    if (tagName != null) {
+      giftCertificateDtos =
+          giftCertificateService.getGiftCertificateListByTagName(
+              giftCertificateDtos, tagName, sort);
+    }
+    return giftCertificateDtos;
+  }
+
+  @GetMapping("/{id}")
+  public GiftCertificateDto get(@PathVariable Long id) {
+    Optional<GiftCertificateDto> optionalGiftCertificateDto = giftCertificateService.getById(id);
+    return optionalGiftCertificateDto.orElseThrow(
+        () -> new EntityNotFoundException("Requested resource not found (id = " + id + ")"));
+  }
+
+  @PostMapping()
+  @ResponseStatus(HttpStatus.OK)
+  public GiftCertificateDto save(@RequestBody GiftCertificateDto giftCertificateDto) {
+    try {
+      Optional<GiftCertificateDto> optionalGiftCertificateDto =
+          giftCertificateService.save(giftCertificateDto);
+      return optionalGiftCertificateDto.orElseThrow(
+          () -> new EntityNotFoundException("Gift certificate didn't add to DB"));
+    } catch (ServiceException e) {
+      throw new InvalidRequestBodyException(e.getMessage());
+    }
+  }
+
+  @PutMapping("/{id}")
+  @ResponseStatus(HttpStatus.OK)
+  public GiftCertificateDto update(
+      @PathVariable long id, @RequestBody GiftCertificateDto giftCertificateDto) {
+    return giftCertificateService
+        .update(id, giftCertificateDto)
+        .orElseThrow(
+            () -> new EntityNotFoundException("Gift certificate with id=" + id + " didn't update"));
+  }
+
+  @DeleteMapping("/{id}")
+  public String delete(@PathVariable long id) {
```

Comment:
@TimurRadko Could you return empty body and `204 No Content` status for `DELETE` endpoints please

Reply:
Yes, of course, I can return an empty body. (fixed)